docs/concurrency_design.md

```diff
@@ -129,3 +129,90 @@ sequenceDiagram
 Some pending definitions:
 
 - The block tracker could eventually be a block cache, and maintain blocks and their state in an ETS that can be accessed easily by other processes.
+
+## Another possible design: per-block processes
+
+Conceptually, the core concurrency issue of the current design is that we use processes to represent steps instead of them representing independent pieces of computation/data. Most of the inter-process communication is call-based and sequential, effectively blocking processes while still adding the overhead of message passing and mailboxes with big objects like blocks and states.
+
+What we need is to have one process per independent task. In our case, our independent unit for is **the block processing**. We need to have a **single process per block**, that takes care of:
+
+- Decompressing
+- Deserializing
+- Saving the block to the DB/caches/fork-tree
+- Validating
+- Performing state transition.
+- Saving the state to the DB/cache
+- At the end of all, updating the global store (this is a very fast operation, contrary to state transition).
+
+With this strategy in mind, the block won’t need to be passed around during its lifetime, and state transitions may happen in parallel without the need to spawn tasks on demand.
+
+There are some tasks that, of course, will require interaction with the outside world:
+
+- When a parent is missing, we need to download it. The block processor may pause until it receives a notification that the parent was downloaded and transitioned. After that, it may continue with its own state transition.
+- When validating it will need to interact with the global fork choice store. This just requires validating a few parameters like the epoch/slot, which is a fast process, so it’s a relatively simple call without passing the full block and without changing the store state.
+- The final step is updating the store, which still doesn’t need to pass around big state and is fast. If additional validations are performed or repeated, it doesn’t represent a big performance hit.
+
+### Happy path diagram
```

Comment:
just FYI, we also have a new abstraction called BeaconChain that should take some of the responsabilty that ForkChoice has today.